cluster-autoscaler/cloudprovider/digitalocean/digitalocean_node_group.go

```diff
@@ -0,0 +1,269 @@
+		if err != nil {
+			return fmt.Errorf("deleting node failed for cluster: %q node pool: %q node: %q: %s",
+				n.clusterID, n.id, nodeID, err)
+		}
+	}
+
+	return nil
+}
+
+// DecreaseTargetSize decreases the target size of the node group. This function
+// doesn't permit to delete any existing node and can be used only to reduce the
+// request for new nodes that have not been yet fulfilled. Delta should be negative.
+// It is assumed that cloud provider will not delete the existing nodes when there
+// is an option to just decrease the target. Implementation required.
+func (n *NodeGroup) DecreaseTargetSize(delta int) error {
+	if delta >= 0 {
+		return fmt.Errorf("delta must be negative, have: %d", delta)
+	}
+
+	targetSize := n.nodePool.Count + delta
+	if targetSize <= n.MinSize() {
+		return fmt.Errorf("size decrease is too small. current: %d desired: %d max: %d",
+			n.nodePool.Count, targetSize, n.MinSize())
+	}
+
+	req := &godo.KubernetesNodePoolUpdateRequest{
+		Count: targetSize,
+	}
+
+	ctx := context.Background()
+	updatedNodePool, _, err := n.client.UpdateNodePool(ctx, n.clusterID, n.id, req)
+	if err != nil {
+		return err
+	}
+
+	if updatedNodePool.Count != targetSize {
+		return fmt.Errorf("couldn't increase size to %d (delta: %d). Current size is: %d",
+			targetSize, delta, updatedNodePool.Count)
+	}
+
+	return nil
+}
+
+// Id returns an unique identifier of the node group.
+func (n *NodeGroup) Id() string {
+	return n.id
+}
+
+// Debug returns a string containing all information regarding this node group.
+func (n *NodeGroup) Debug() string {
+	return fmt.Sprintf("cluster ID: %s (min:%d max:%d)", n.Id(), n.MinSize(), n.MaxSize())
+}
+
+// Nodes returns a list of all nodes that belong to this node group.  It is
+// required that Instance objects returned by this method have Id field set.
+// Other fields are optional.
+func (n *NodeGroup) Nodes() ([]cloudprovider.Instance, error) {
+	if n.nodePool == nil {
+		return nil, errors.New("node pool instance is not created")
+	}
+	return toInstances(n.nodePool.Nodes), nil
+}
+
+// TemplateNodeInfo returns a schedulernodeinfo.NodeInfo structure of an empty
+// (as if just started) node. This will be used in scale-up simulations to
+// predict what would a new node look like if a node group was expanded. The
+// returned NodeInfo is expected to have a fully populated Node object, with
+// all of the labels, capacity and allocatable information as well as all pods
+// that are started on the node by default, using manifest (most likely only
+// kube-proxy). Implementation optional.
+func (n *NodeGroup) TemplateNodeInfo() (*schedulernodeinfo.NodeInfo, error) {
+	return nil, cloudprovider.ErrNotImplemented
+}
+
+// Exist checks if the node group really exists on the cloud provider side.
+// Allows to tell the theoretical node group from the real one. Implementation
+// required.
+func (n *NodeGroup) Exist() bool {
+	return n.nodePool != nil
```

Comment:
Wondering if this should actually call the provider cause a node pool could be deleted between calls to `Refresh`?

Reply:
Looks like the other providers don't do this so maybe not

Reply:
@andrewsykim that was the case for me when I wrote this initially. However upon using this and checking the code, seems like `Refresh()` is called every `--scan-interval` and then all the methods in this file are called based on when they are need or not.  The cluster-autoscaler basically reconciles every `--scan-interval` (default 10 seconds) therefore we should be fairly stable for race conditions. 

The other issue is, using `Refresh()` in these methods triggers the rate limit on the public API of DigitalOcean, because the cluster-autoscaler aggressively calls these during each reconcile period. Hence the most efficient and performant  way (based on my observation) is to use an internal cache that is populated (therefore invalidated) each reconcile period. Which is defined by the `--scan-interval` period.

Reply:
That seems reasonable to me